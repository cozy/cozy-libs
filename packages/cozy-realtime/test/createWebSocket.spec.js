import { Server } from 'mock-socket'

import __RewireAPI__, { createWebSocket, getSocket } from '../src/legacy'

const MOCK_SERVER_DOMAIN = 'localhost:8880'

const REALTIME_URL = `ws://${MOCK_SERVER_DOMAIN}/realtime/`
const REALTIME_URL_SECURE = `wss://${MOCK_SERVER_DOMAIN}/realtime/`

let server
let mockSubscribe = jest.fn()
jest.useFakeTimers() // mock-socket use timers to delay onopen call
describe('(cozy-realtime) createWebSocket: ', () => {
  beforeEach(() => {
    jest.clearAllMocks()
    jest.resetModules()
    __RewireAPI__.__Rewire__('subscribeWhenReady', mockSubscribe)
    __RewireAPI__.__Rewire__('socketPromise', null)
    server = new Server(REALTIME_URL)
  })

  afterEach(() => {
    jest.runAllTimers()
    __RewireAPI__.__ResetDependency__('subscribeWhenReady')
    __RewireAPI__.__ResetDependency__('socketPromise')
    server.stop()
  })

  it('socket should create a global socket promise with provided domain and secure option', async () => {
    const mockConfig = {
      domain: MOCK_SERVER_DOMAIN,
      secure: false,
      token: 'blablablatoken'
    }

    createWebSocket(mockConfig, jest.fn(), jest.fn(), 10, 2000)
    jest.runAllTimers()
    expect(await getSocket()).toMatchSnapshot()
  })

  it('socket should create a global socket handling wss', async () => {
    server.stop()
    server = new Server(REALTIME_URL_SECURE)
    const mockConfig = {
      domain: MOCK_SERVER_DOMAIN,
      secure: true,
      token: 'blablablatoken'
    }

    createWebSocket(mockConfig, jest.fn(), jest.fn(), 10, 2000)
    jest.runAllTimers()
    expect(await getSocket()).toMatchSnapshot()
  })

  it('socket should throw error if no url or domain provided', () => {
    const mockConfig = {
      secure: false,
      token: 'blablablatoken'
    }

    expect(() => {
      createWebSocket(mockConfig, jest.fn(), jest.fn(), 10, 2000)
      jest.runAllTimers()
    }).toThrowErrorMatchingSnapshot()
  })

  it('socket should throw error if wrong url format provided', () => {
    const mockConfig = {
      url: 'blable.bla:blabla',
      secure: false,
      token: 'blablablatoken'
    }

    expect(() => {
      createWebSocket(mockConfig, jest.fn(), jest.fn(), 10, 2000)
      jest.runAllTimers()
    }).toThrowErrorMatchingSnapshot()
  })

  it('socket should throw error if wrong url type provided', () => {
    const mockConfig = {
      url: () => {},
      secure: false,
      token: 'blablablatoken'
    }

    expect(() => {
      createWebSocket(mockConfig, jest.fn(), jest.fn(), 10, 2000)
      jest.runAllTimers()
    }).toThrowErrorMatchingSnapshot()
  })

  it('socket should handle authenticating on socket open', async () => {
    const mockConfig = {
      domain: MOCK_SERVER_DOMAIN,
      secure: false,
      token: 'blablablatoken'
    }
    const sentMessages = []
    server.on('connection', socket => {
      socket.on('message', data => sentMessages.push(data))
    })

    createWebSocket(mockConfig, jest.fn(), jest.fn(), 10, 2000)
    jest.runAllTimers()
    expect(sentMessages.length).toBe(1)
    expect(JSON.parse(sentMessages[0]).payload).toBe(mockConfig.token)
    expect(JSON.parse(sentMessages[0])).toMatchSnapshot()
  })

  it('socket should warn errors on socket errors', () => {
    const mockConfig = {
      domain: MOCK_SERVER_DOMAIN,
      secure: false,
      token: 'blablablatoken'
    }
    console.error = jest.fn()

    createWebSocket(mockConfig, jest.fn(), jest.fn(), 10, 2000)
    jest.runAllTimers()
    // simulate onerror
    expect(() => {
      server.simulate('error')
    }).not.toThrowError()
    expect(console.error.mock.calls.length).toBe(1)
    expect(console.error.mock.calls[0][0]).toMatchSnapshot()
    console.error.mockRestore()
  })

  it('socket should handle message', () => {
    const mockConfig = {
      domain: MOCK_SERVER_DOMAIN,
      secure: false,
      token: 'blablablatoken'
    }
    const onMessageMock = jest.fn()

    createWebSocket(mockConfig, onMessageMock, jest.fn(), 10, 2000)
    jest.runAllTimers()
    // simulate a message
    server.emit('message', 'a server message to socket')
    expect(onMessageMock.mock.calls.length).toBe(1)
    onMessageMock.mock.calls[0][0].timeStamp = 0 // reset timestamp for snapshot
    expect(onMessageMock.mock.calls[0][0]).toMatchSnapshot()
  })

  it('socket should handle closing server', () => {
    const mockConfig = {
      domain: MOCK_SERVER_DOMAIN,
      secure: false,
      token: 'blablablatoken'
    }
    const onCloseMock = jest.fn()
    jest.spyOn(window, 'removeEventListener')

    createWebSocket(mockConfig, jest.fn(), onCloseMock, 10, 2000)
    jest.runAllTimers()
    server.close()
    expect(onCloseMock.mock.calls.length).toBe(1)
    expect(window.removeEventListener.mock.calls.length).toBe(1)
    onCloseMock.mock.calls[0][0].timeStamp = 0 // reset timestamp for snapshot
    expect(onCloseMock.mock.calls[0][0]).toMatchSnapshot()
  })

  it('socket should handle closing server even if no onclose function provided', () => {
    const mockConfig = {
      domain: MOCK_SERVER_DOMAIN,
      secure: false,
      token: 'blablablatoken'
    }

    createWebSocket(mockConfig, jest.fn(), null, 10, 2000)
    jest.runAllTimers()
    server.close()
    expect(window.removeEventListener.mock.calls.length).toBe(1)
  })

  it('socket should close the socket on unloading window', async () => {
    const mockConfig = {
      domain: MOCK_SERVER_DOMAIN,
      secure: false,
      token: 'blablablatoken'
    }

    createWebSocket(mockConfig, jest.fn(), jest.fn(), 10, 2000)
    jest.runAllTimers()
    const cozySocket = await getSocket()
    expect(cozySocket.readyState).toBe(1) // OPENED
    window.dispatchEvent(new Event('beforeunload'))
    expect(cozySocket.readyState).toBe(2) // CLOSING
    jest.runAllTimers()
    expect(cozySocket.readyState).toBe(3) // CLOSED
  })
})
