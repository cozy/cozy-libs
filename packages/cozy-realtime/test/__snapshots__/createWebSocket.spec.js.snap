// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(cozy-realtime) createWebSocket:  socket should create a global socket handling wss 1`] = `
WebSocket {
  "binaryType": "blob",
  "listeners": Object {
    "close": Array [
      [Function],
    ],
    "error": Array [
      [Function],
    ],
    "message": Array [
      [MockFunction],
    ],
    "open": Array [
      [Function],
    ],
  },
  "protocol": "io.cozy.websocket",
  "readyState": 1,
  "url": "wss://localhost:8880/realtime/",
}
`;

exports[`(cozy-realtime) createWebSocket:  socket should create a global socket promise with provided domain and secure option 1`] = `
WebSocket {
  "binaryType": "blob",
  "listeners": Object {
    "close": Array [
      [Function],
    ],
    "error": Array [
      [Function],
    ],
    "message": Array [
      [MockFunction],
    ],
    "open": Array [
      [Function],
    ],
  },
  "protocol": "io.cozy.websocket",
  "readyState": 1,
  "url": "ws://localhost:8880/realtime/",
}
`;

exports[`(cozy-realtime) createWebSocket:  socket should handle authenticating on socket open 1`] = `
Object {
  "method": "AUTH",
  "payload": "blablablatoken",
}
`;

exports[`(cozy-realtime) createWebSocket:  socket should handle closing server 1`] = `
CloseEvent {
  "bubbles": false,
  "cancelBubble": false,
  "cancelable": false,
  "code": 1000,
  "currentTarget": WebSocket {
    "binaryType": "blob",
    "listeners": Object {
      "close": Array [
        [Function],
      ],
      "error": Array [
        [Function],
      ],
      "message": Array [
        [MockFunction],
      ],
      "open": Array [
        [Function],
      ],
    },
    "protocol": "io.cozy.websocket",
    "readyState": undefined,
    "url": "ws://localhost:8880/realtime/",
  },
  "defaultPrevented": false,
  "eventPhase": 0,
  "isTrusted": false,
  "reason": "",
  "returnValue": true,
  "srcElement": WebSocket {
    "binaryType": "blob",
    "listeners": Object {
      "close": Array [
        [Function],
      ],
      "error": Array [
        [Function],
      ],
      "message": Array [
        [MockFunction],
      ],
      "open": Array [
        [Function],
      ],
    },
    "protocol": "io.cozy.websocket",
    "readyState": undefined,
    "url": "ws://localhost:8880/realtime/",
  },
  "target": WebSocket {
    "binaryType": "blob",
    "listeners": Object {
      "close": Array [
        [Function],
      ],
      "error": Array [
        [Function],
      ],
      "message": Array [
        [MockFunction],
      ],
      "open": Array [
        [Function],
      ],
    },
    "protocol": "io.cozy.websocket",
    "readyState": undefined,
    "url": "ws://localhost:8880/realtime/",
  },
  "timeStamp": 0,
  "type": "close",
  "wasClean": true,
}
`;

exports[`(cozy-realtime) createWebSocket:  socket should handle message 1`] = `
MessageEvent {
  "bubbles": false,
  "cancelable": false,
  "canncelBubble": false,
  "currentTarget": WebSocket {
    "binaryType": "blob",
    "listeners": Object {
      "close": Array [
        [Function],
      ],
      "error": Array [
        [Function],
      ],
      "message": Array [
        [MockFunction] {
          "calls": Array [
            Array [
              [Circular],
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      ],
      "open": Array [
        [Function],
      ],
    },
    "protocol": "io.cozy.websocket",
    "readyState": 1,
    "url": "ws://localhost:8880/realtime/",
  },
  "data": "a server message to socket",
  "defaultPrevented": false,
  "eventPhase": 0,
  "isTrusted": false,
  "lastEventId": "",
  "origin": "ws://localhost:8880/realtime/",
  "ports": null,
  "returnValue": true,
  "srcElement": WebSocket {
    "binaryType": "blob",
    "listeners": Object {
      "close": Array [
        [Function],
      ],
      "error": Array [
        [Function],
      ],
      "message": Array [
        [MockFunction] {
          "calls": Array [
            Array [
              [Circular],
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      ],
      "open": Array [
        [Function],
      ],
    },
    "protocol": "io.cozy.websocket",
    "readyState": 1,
    "url": "ws://localhost:8880/realtime/",
  },
  "target": WebSocket {
    "binaryType": "blob",
    "listeners": Object {
      "close": Array [
        [Function],
      ],
      "error": Array [
        [Function],
      ],
      "message": Array [
        [MockFunction] {
          "calls": Array [
            Array [
              [Circular],
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      ],
      "open": Array [
        [Function],
      ],
    },
    "protocol": "io.cozy.websocket",
    "readyState": 1,
    "url": "ws://localhost:8880/realtime/",
  },
  "timeStamp": 0,
  "type": "message",
}
`;

exports[`(cozy-realtime) createWebSocket:  socket should throw error if no url or domain provided 1`] = `"domain is required if no attribute url are provider.."`;

exports[`(cozy-realtime) createWebSocket:  socket should throw error if wrong url format provided 1`] = `"Unable to detect domain"`;

exports[`(cozy-realtime) createWebSocket:  socket should throw error if wrong url type provided 1`] = `"url should be an URL."`;

exports[`(cozy-realtime) createWebSocket:  socket should warn errors on socket errors 1`] = `"WebSocket error: undefined"`;
